codegeex.el

```diff
@@ -125,17 +118,105 @@ buffer."
          (name-without-mode (replace-regexp-in-string "-mode" "" name)))
     name-without-mode))
 
+(defun codegeex--plist-get (plist &rest keys)
+  "Browse PLIST object for each KEYS and return the element
+where it stop.
+KEYS can be either numbers or properties symbols"
+  (let ((res json-result))
+    (dolist (key keys)
+      (cond
+       ((symbolp key)
+        (setq res (plist-get res key)))
+       ((numberp key)
+        (setq res (nth key res)))
+       (t
+        (error "Key '%s' format not supported" key))))
+    res))
+
+(defvar-local codegeex--completion-cache nil)
+(defvar-local codegeex--completion-idx 0)
+
 ;;;###autoload
-(defun codegeex-buffer-completion ()
-  "CodeGeeX buffer completion.
+(defun codegeex-complete ()
+  "Get completion at point, showing the completion in an overlay"
+  (interactive)
+  (setq codegeex--completion-cache nil)
+  (codegeex-completion--get-completion
+   (lambda (json-result)
+     (let* ((completions
+             (cl-remove-if
+              (lambda (e)
+                (string= "" (string-trim e)))
+              (cl-remove-duplicates
+               (codegeex--plist-get
+                json-result :result :output :code))))
+           (completion (if (seq-empty-p completions) nil (seq-elt completions 0))))
+       (setq codegeex--completion-cache completions)
+       (if completion
+           (codegeex-completion--show-completion completion)
+         (message "No completion available"))))))
 
-The completion result will be put in the position of cursor directly."
+;;;###autoload
+(defun codegeex-complete-at-point ()
+  "Get first completion proposed and insert it at point
+Can be used without having to be in `codegeex-mode'"
   (interactive)
-  (message "CodeGeeX completing")
-  (let ((prefix (buffer-substring (point-min) (point)))
-        (suffix (buffer-substring (point) (point-max))))
-    (codegeex-completion-invoke prefix suffix (codegeex-language)))
-  nil)
+  (codegeex-completion--get-completion
+   (lambda (json-result)
+     (let ((completion (codegeex--plist-get
+                        json-result :result :output :code 0)))
+       (if completion
+         (insert completion)
+       (message "No completion available"))))))
+
+;;;###autoload
+(defun codegeex-accept-completion (&optional transform-fn)
+  "Accept completion. Return t if there is a completion.
+Use TRANSFORM-FN to transform completion if provided."
+  (interactive)
+  (when (codegeex--overlay-visible)
+    (let* ((completion (overlay-get codegeex--overlay 'completion))
+           (start (overlay-get codegeex--overlay 'start))
+           (end (codegeex--overlay-end codegeex--overlay))
+           (uuid (overlay-get codegeex--overlay 'uuid))
+           (t-completion (funcall (or transform-fn #'identity) completion)))
+      (codegeex-clear-overlay)
+      (if (eq major-mode 'vterm-mode)
+          (progn
+            (vterm-delete-region start end)
```

Comment:
`codegeex-accept-completion` `codegeex-next-completion` and `codegeex-previous-completion` are use to handle the overlaid completion proposition fetched. They can be binded for convenience by the user